File: HisabCore/Sources/HisabCore/IDFCXLSXParser.swift
import Foundation

/// Parser for IDFC FIRST Bank .xlsx statements. Unlike the PDF rendition, the sheet
/// has explicit Debit/Credit numeric columns; the running Balance column is still
/// verified so a misread row throws instead of miscounting. Ref-less rows get the
/// same balance-keyed synthetic reference as the PDF parser, so importing both
/// renditions of one statement dedups completely.
public struct IDFCXLSXParser: StatementParser {
    public let source = Source.idfc

    public init() {}

    private static var periodRegex: Regex<(Substring, Substring, Substring)> {
        /^(\d{2}-[A-Za-z]{3}-\d{4}) TO (\d{2}-[A-Za-z]{3}-\d{4})$/
    }

    public func canParse(data: Data, filename: String) -> Bool {
        guard data.starts(with: [0x50, 0x4B]) else { return false }
        guard let workbook = try? XLSXReader.read(data: data) else { return false }
        return workbook.sharedStrings.contains("STATEMENT OF ACCOUNT")
            && workbook.sharedStrings.contains { $0.hasPrefix("IDFB") }
    }

    public func parse(data: Data, password: String?) throws -> ParsedDocument {
        let workbook: XLSXReader.Workbook
        do {
            workbook = try XLSXReader.read(data: data)
        } catch {
            throw ParseError.unrecognizedFormat
        }

        guard let rows = workbook.sheets.values.first(where: { sheet in
            sheet.contains { $0.values.contains("Transaction Date") }
        }) else { throw ParseError.unrecognizedFormat }
        guard let headerIndex = rows.firstIndex(where: { $0.values.contains("Transaction Date") }) else {
            throw ParseError.unrecognizedFormat
        }
        let header = rows[headerIndex]
        func column(_ title: String) -> String? { header.first { $0.value == title }?.key }
        guard let dateCol = column("Transaction Date"),
              let particularsCol = column("Particulars"),
              let debitCol = column("Debit"),
              let creditCol = column("Credit"),
              let balanceCol = column("Balance") else {
            throw ParseError.unrecognizedFormat
        }

        let dateFormatter = DateFormatter()
        dateFormatter.calendar = YearMonth.istCalendar
        dateFormatter.timeZone = YearMonth.istCalendar.timeZone
        dateFormatter.locale = Locale(identifier: "en_US_POSIX")
        dateFormatter.dateFormat = "dd-MMM-yyyy"

        var previousBalance: Int64?
        var transactions: [ParsedTransaction] = []

        for row in rows.dropFirst(headerIndex + 1) {
            guard let dateText = row[dateCol], let date = dateFormatter.date(from: dateText),
                  let particulars = row[particularsCol],
                  let balanceText = row[balanceCol],
                  let balance = Self.paise(fromNumericCell: balanceText) else { continue }

            let debit = row[debitCol].flatMap(Self.paise(fromNumericCell:))
            let credit = row[creditCol].flatMap(Self.paise(fromNumericCell:))
            let direction: Direction
            let amount: Int64
            switch (debit, credit) {
            case (let d?, nil): direction = .debit; amount = d
            case (nil, let c?): direction = .credit; amount = c
            default: throw ParseError.malformedRow(0, particulars)
            }

            if let previous = previousBalance {
                let expected = direction == .debit ? previous - amount : previous + amount
                guard expected == balance else {
                    throw ParseError.malformedRow(0, "balance chain break at: \(particulars)")
                }
            }
            previousBalance = balance

            let (counterparty, reference) = IDFCStatementText.extract(from: particulars)
            transactions.append(ParsedTransaction(
                date: date, amountPaise: amount, direction: direction,
                counterparty: counterparty,
                reference: reference ?? IDFCStatementText.syntheticReference(
                    balancePaise: balance, date: date, amountPaise: amount),
                narration: particulars))
        }
        guard !transactions.isEmpty else { throw ParseError.empty }

        var declaredPeriod: DatePeriod?
        for text in workbook.sharedStrings {
            guard let match = text.wholeMatch(of: Self.periodRegex),
                  let start = dateFormatter.date(from: String(match.1)),
                  let end = dateFormatter.date(from: String(match.2)) else { continue }
            declaredPeriod = DatePeriod(start: start, end: end)
            break
        }
        return ParsedDocument(source: .idfc, declaredPeriod: declaredPeriod, transactions: transactions)
    }

    /// "303594.0" / "1417816.76" / "1999" -> paise, exact via Decimal.
    static func paise(fromNumericCell raw: String) -> Int64? {
        let text = raw.trimmingCharacters(in: .whitespaces)
        guard !text.isEmpty, let value = Decimal(string: text, locale: Locale(identifier: "en_US_POSIX")) else {
            return nil
        }
        var scaled = value * 100
        var rounded = Decimal()
        NSDecimalRound(&rounded, &scaled, 0, .plain)
        return (rounded as NSDecimalNumber).int64Value
    }
}

File: hisab_flutter/packages/hisab_core/lib/src/parsers/idfc_xlsx_parser.dart
/// IDFC FIRST Bank .xlsx statements. Port of IDFCXLSXParser.swift — explicit
/// Debit/Credit columns, running Balance verified, balance-keyed synthetic
/// refs for ref-less rows.
library;

import '../domain.dart';
import '../generic_bank/statement_date.dart';
import '../generic_bank/synthetic_ref.dart';
import '../generic_bank/xlsx_reader.dart';
import 'idfc_common.dart';

class IdfcXlsxParser implements StatementParser {
  @override
  Source get source => Source.idfc;

  const IdfcXlsxParser();

  @override
  bool canParse(List<int> data, String filename) {
    if (data.length < 2 || data[0] != 0x50 || data[1] != 0x4B) return false;
    final workbook = XlsxReader.read(data);
    if (workbook == null) return false;
    return workbook.sharedStrings.contains('STATEMENT OF ACCOUNT') &&
        workbook.sharedStrings.any((s) => s.startsWith('IDFB'));
  }

  @override
  ParsedDocument parse(List<int> data, {String? password}) {
    final workbook = XlsxReader.read(data);
    if (workbook == null) throw const UnrecognizedFormatException();

    List<Map<String, String>>? rows;
    for (final sheet in workbook.sheets.values) {
      if (sheet.any((r) => r.containsValue('Transaction Date'))) {
        rows = sheet;
        break;
      }
    }
    if (rows == null) throw const UnrecognizedFormatException();
    final headerIndex =
        rows.indexWhere((r) => r.containsValue('Transaction Date'));
    if (headerIndex < 0) throw const UnrecognizedFormatException();
    final header = rows[headerIndex];
    String? column(String title) {
      for (final entry in header.entries) {
        if (entry.value == title) return entry.key;
      }
      return null;
    }

    final dateCol = column('Transaction Date');
    final particularsCol = column('Particulars');
    final debitCol = column('Debit');
    final creditCol = column('Credit');
    final balanceCol = column('Balance');
    if (dateCol == null ||
        particularsCol == null ||
        debitCol == null ||
        creditCol == null ||
        balanceCol == null) {
      throw const UnrecognizedFormatException();
    }

    int? previousBalance;
    final transactions = <ParsedTransaction>[];

    for (final row in rows.skip(headerIndex + 1)) {
      final dateText = row[dateCol];
      final date = dateText == null
          ? null
          : StatementDate.tryParse(dateText, 'dd-MMM-yyyy');
      final particulars = row[particularsCol];
      final balanceText = row[balanceCol];
      final balance = balanceText == null ? null : _paise(balanceText);
      if (date == null || particulars == null || balance == null) continue;

      final debit = row[debitCol] == null ? null : _paise(row[debitCol]!);
      final credit = row[creditCol] == null ? null : _paise(row[creditCol]!);
      Direction direction;
      int amount;
      if (debit != null && credit == null) {
        direction = Direction.debit;
        amount = debit;
      } else if (debit == null && credit != null) {
        direction = Direction.credit;
        amount = credit;
      } else {
        throw MalformedRowException(0, particulars);
      }

      final previous = previousBalance;
      if (previous != null) {
        final expected =
            direction == Direction.debit ? previous - amount : previous + amount;
        if (expected != balance) {
          throw MalformedRowException(0, 'balance chain break at: $particulars');
        }
      }
      previousBalance = balance;

      final (counterparty, reference) = IdfcStatementText.extract(particulars);
      transactions.add(ParsedTransaction(
        date: date,
        amountPaise: amount,
        direction: direction,
        counterparty: counterparty,
        reference: reference ??
            SyntheticRef.make(
                balancePaise: balance, date: date, amountPaise: amount),
        narration: particulars,
      ));
    }
    if (transactions.isEmpty) throw const EmptyDocumentException();

    DatePeriod? declaredPeriod;
    final periodRegex =
        RegExp(r'^(\d{2}-[A-Za-z]{3}-\d{4}) TO (\d{2}-[A-Za-z]{3}-\d{4})$');
    for (final text in workbook.sharedStrings) {
      final match = periodRegex.firstMatch(text);
      if (match == null) continue;
      final start = StatementDate.tryParse(match.group(1)!, 'dd-MMM-yyyy');
      final end = StatementDate.tryParse(match.group(2)!, 'dd-MMM-yyyy');
      if (start == null || end == null) continue;
      declaredPeriod = DatePeriod(start, end);
      break;
    }
    return ParsedDocument(
        source: Source.idfc,
        declaredPeriod: declaredPeriod,
        transactions: transactions);
  }

  /// "303594.0" / "1417816.76" / "1999" → paise, exact. Mirrors the Swift
  /// Decimal-based conversion for raw xlsx numeric cells.
  static int? _paise(String raw) {
    final text = raw.trim();
    if (text.isEmpty) return null;
    final negative = text.startsWith('-');
    final unsigned = negative ? text.substring(1) : text;
    final parts = unsigned.split('.');
    if (parts.length > 2 || parts[0].isEmpty) return null;
    final rupees = int.tryParse(parts[0]);
    if (rupees == null) return null;
    var fraction = 0;
    if (parts.length == 2 && parts[1].isNotEmpty) {
      var digits = parts[1];
      if (digits.length == 1) digits = '${digits}0';
      if (digits.length > 2) {
        // Round half-even like NSDecimalRound(.plain)? Statement cells carry
        // at most 2 decimals; longer means float noise — round half up.
        final head = int.tryParse(digits.substring(0, 2));
        final next = int.tryParse(digits.substring(2, 3));
        if (head == null || next == null) return null;
        fraction = next >= 5 ? head + 1 : head;
      } else {
        final value = int.tryParse(digits);
        if (value == null) return null;
        fraction = value;
      }
    }
    final paise = rupees * 100 + fraction;
    return negative ? -paise : paise;
  }
}
